### DISCUSSION TITLE
gogs server key

### DISCUSSION THREAD

**Discussion by waldemara** [gogs/gogs#7063] [General]
Recently, I am getting "Unable to negotiate with x.x.x.x port 22: no matching host key type found. Their offer: ssh-rsa" when connecting to gogs server. This can be temporarily fixed by adding 'HostkeyAlgorithms +ssh-rsa' to ssh config file but real problem is that gogs is using obsolete keys.
How do I change gogs server keys?  This is for the embedded ssh server.

**Comment by waldemara** [comment]
Hmm. Now it gets worse. I figured maybe I upgrade to 0.12.9 and the server will do the right thing upgrading keys. Not so. It just fails:
`[gogs.io/gogs/internal/ssh/ssh.go:151 func1()] SSH: Error on handshaking: ssh: no common algorithm for host key; client offered: [rsa-sha2-512-cert-v01@openssh.com rsa-sha2-256-cert-v01@openssh.com rsa-sha2-512 rsa-sha2-256 ssh-ed25519-cert-v01@openssh.com ecdsa-sha2-nistp256-cert-v01@openssh.com ecdsa-sha2-nistp384-cert-v01@openssh.com ecdsa-sha2-nistp521-cert-v01@openssh.com sk-ssh-ed25519-cert-v01@openssh.com sk-ecdsa-sha2-nistp256-cert-v01@openssh.com ssh-ed25519 ecdsa-sha2-nistp256 ecdsa-sha2-nistp384 ecdsa-sha2-nistp521 sk-ssh-ed25519@openssh.com sk-ecdsa-sha2-nistp256@openssh.com], server offered: [ssh-rsa]`
I guess, I'll try to hack it but gogs should be able to upgrade its own keys. The user, or admin rather, is given no means to do that. I think this is worth raising as an issue.

**Reply by unknwon** [reply]
Here is the relevant part if you want to hack:

https://github.com/gogs/gogs/blob/97ccb365ecc8312a07f561792be4075e43c43d96/internal/ssh/ssh.go#L181-L191

You can also follow on https://github.com/gogs/gogs/issues/6638
